From 5e06c89c7a58a22c020056eb92e81f3e2855f80f Mon Sep 17 00:00:00 2001
From: Zhang Yanmin <yanmin.zhang@intel.com>
Date: Tue, 9 Jun 2015 08:15:04 +0800
Subject: [PATCH] dwc3: call disconnect before stopping dwc3

We hit adb offline issues on mobiles frequently.
Basically, when android disables usb, dwc3_gadget_pullup
is called to stop dwc3. After dwc3 stops, callback
disconnect is called to disconnect functions.

The sequence is incorrect. Kernel has to call disconnect
firstly, then stop dwc3.

The patch fixes it by reversing them.

Change-Id: Id9e5d7aaead944067949efb269cd1a2370e8c4d5
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-33274
Signed-off-by: Zhang Yanmin <yanmin.zhang@intel.com>
---
 drivers/usb/dwc3/gadget.c | 23 +++++++++++------------
 1 file changed, 11 insertions(+), 12 deletions(-)

diff --git a/drivers/usb/dwc3/gadget.c b/drivers/usb/dwc3/gadget.c
index 1e71d2ef7aa0..b187182fb55d 100644
--- a/drivers/usb/dwc3/gadget.c
+++ b/drivers/usb/dwc3/gadget.c
@@ -1588,20 +1588,13 @@ static int dwc3_gadget_pullup(struct usb_gadget *g, int is_on)
 
 	pm_runtime_get_sync(dwc->dev);
 
-	spin_lock_irqsave(&dwc->lock, flags);
-	ret = dwc3_gadget_run_stop(dwc, is_on);
-	spin_unlock_irqrestore(&dwc->lock, flags);
-
-	pm_runtime_put_autosuspend(dwc->dev);
-
 	/**
-	 * WORKAROUND: The dwc3 controller doesn't issue disconnect
-	 * event after pullups is disabled. Thus gadget dissconnect callback
-	 * is not invoked when do soft dissconnect. Call dissconnect here
-	 * to workaround this issue.
+	 * WORKAROUND: The dwc3 controller doesn't issue disconnect.
+	 * Thus gadget dissconnect callback is not invoked when do
+	 * soft dissconnect. Call dissconnect here to workaround
+	 * this issue.
 	 */
-	if (!ret && !is_on
-		&& dwc->gadget_driver && dwc->gadget_driver->disconnect) {
+	if (!is_on && dwc->gadget_driver && dwc->gadget_driver->disconnect) {
 		if (dwc->gadget.speed != USB_SPEED_UNKNOWN) {
 			dev_dbg(dwc->dev, "fake dissconnect event\n");
 			irq = platform_get_irq(to_platform_device(dwc->dev), 0);
@@ -1612,6 +1605,12 @@ static int dwc3_gadget_pullup(struct usb_gadget *g, int is_on)
 		}
 	}
 
+	spin_lock_irqsave(&dwc->lock, flags);
+	ret = dwc3_gadget_run_stop(dwc, is_on);
+	spin_unlock_irqrestore(&dwc->lock, flags);
+
+	pm_runtime_put_autosuspend(dwc->dev);
+
 	return ret;
 }
 
-- 
1.9.1

